Tests/DistributedClusterTests/Plugins/ClusterSingleton/ClusterSingletonPluginTests.swift

```diff
@@ -59,6 +59,48 @@ final class ClusterSingletonPluginTests: SingleClusterSystemXCTestCase {
         // if this were true we would have crashed by a duplicate name already, but let's make sure:
         singletonID.shouldNotEqual(greeterID)
     }
+    
+    func test_plugin_hooks() async throws {
+        let actorId = "actorHookId"
+        let hookFulfillment = self.expectation(description: "actor-hook")
+        let plugin = TestClusterHookPlugin { actor in
+            /// There are multiple internal actors fired, we only checking for `ActorWithId`
+            guard let actor = actor as? ActorWithId else { return }
+            let id = try? await actor.getId()
+            XCTAssertEqual(id, actorId, "Expected \(actorId) as an id")
+            hookFulfillment.fulfill()
+        }
+        let testNode = await setUpNode("test-hook") { settings in
+            settings.enabled = false
+            settings += plugin
+        }
+
+        let _ = ActorWithId(actorSystem: testNode, id: actorId)
+        await fulfillment(of: [hookFulfillment])
+    }
+    
+    actor TestClusterHookPlugin: _Plugin, PluginActorLifecycleHook {
+        nonisolated var key: Key { "$testClusterHook" }
+        
+        let onActorReady: (any DistributedActor) async throws -> ()
+        
+        init(
+            onActorReady: @escaping (any DistributedActor) async throws -> Void
+        ) {
+            self.onActorReady = onActorReady
+        }
+        
+        nonisolated func actorReady<Act>(_ actor: Act) where Act: DistributedActor, Act.ID == DistributedCluster.ClusterSystem.ActorID {
+            Task { try await self.onActorReady(actor) }
```

Comment:
probably make use of a lock inside here and make it a class to avoid the racy nature of Task {}

Reply:
This is why I was worried about sync functions in protocol. Working with actors you'll mostly land in `async` anyway or your plugin could be an actor. And then you'll have `onisolated func actorReady` with `Task` or some later stage.
Which is also ok though if you handle it inside your plugin I guess 🤔

Reply:
I know this is a very old discussion, but to clarify: the actor ready must be synchronous because we support synchronous init() in actors.